### ISSUE TITLE
Improvement: RAG Search

### ISSUE THREAD

**Issue by rmusser01** [rmusser01/tldw#48]
lol.

`Incase you're curious we use named entity recognition models to extract key words / phrases. Then use bm25 + vector search to identify the top results! `
Collections of Papers
	- https://github.com/hymie122/RAG-Survey
	- https://github.com/aishwaryanr/awesome-generative-ai-guide/blob/main/research_updates/rag_research_table.md

https://www.llamaindex.ai/blog/using-llamaindex-and-llamafile-to-build-a-local-private-research-assistant

Unsorted RAG Techniques
- https://arxiv.org/abs/2407.00320
- https://arxiv.org/abs/2409.01666
- https://arxiv.org/pdf/2409.05591
- https://arxiv.org/html/2409.12941v1#S5
- https://github.com/declare-lab/trust-align
- https://arxiv.org/abs/2407.10670
- https://arxiv.org/abs/2409.12941
- https://selfrag.github.io/
LongRAG
	https://github.com/TIGER-AI-Lab/LongRAG
- https://arxiv.org/abs/2402.16406
- https://huggingface.co/papers/2407.14482
- https://huggingface.co/papers/2406.12824
Speculative RAG
https://arxiv.org/abs/2407.08223
Dense X retrieval
https://medium.com/papers-i-found/e14-dense-x-retrieval-d340d20188d3
https://arxiv.org/abs/2312.06648
Speculative RAG
	https://arxiv.org/abs/2407.08223
- https://arxiv.org/abs/2408.13533
- https://arxiv.org/abs/2409.01666
- 
Spreadsheets
https://huggingface.co/papers/2407.09025

RAPTOR
https://arxiv.org/html/2401.18059v1

RAG Techniques Implemented
	https://github.com/ShishirPatil/gorilla/tree/main/raft
	https://arxiv.org/html/2408.06941v1
	https://github.com/NirDiamant/RAG_Techniques
	https://github.com/NirDiamant/Controllable-RAG-Agent
	https://github.com/EdwardDali/erag
	https://github.com/gusye1234/nano-graphrag
https://github.com/NirDiamant/RAG_Techniques

Links:

https://www.youtube.com/watch?v=fWC4VxolWAk
https://www.youtube.com/watch?v=bNqSRNMgwhQ
https://www.youtube.com/watch?v=Balro-DxFyk&list=PLwPYSl1MQp4FpIzn48ypesKYzLvUBQpPF&index=4
https://www.youtube.com/watch?v=vX3A96_F3FU
https://www.youtube.com/watch?v=viZrOnJclY0
https://www.youtube.com/watch?v=7h6uDsfD7bg

https://www.reddit.com/r/LangChain/comments/1dzfp48/agent_retrieval_how_we_almost_always_find_the/
https://www.reddit.com/r/LangChain/comments/1dtr49t/agent_rag_parallel_quotes_how_we_built_rag_on/
https://medium.com/the-ai-forum/rag-on-complex-pdf-using-llamaparse-langchain-and-groq-5b132bd1f9f3

https://www.turingpost.com/p/rag-tutorials-webinars-workshops

https://x.com/hrishioa/status/1735711619588690382
https://x.com/hrishioa/status/1745835962108985737

RAG 101
https://www.youtube.com/watch?v=nc0BupOkrhI
https://arxiv.org/abs/2401.08406

101
https://towardsdatascience.com/how-i-turned-my-companys-docs-into-a-searchable-database-with-openai-4f2d34bd8736

201
https://www.youtube.com/watch?v=0nA5QG3087g
https://d-star.ai/solving-the-out-of-context-chunk-problem-for-rag
https://medium.com/@cdg2718/why-your-rag-doesnt-work-9755726dd1e9
https://www.cazton.com/blogs/technical/advanced-rag-techniques
https://medium.com/@krtarunsingh/advanced-rag-techniques-unlocking-the-next-level-040c205b95bc
https://pub.towardsai.net/advanced-rag-techniques-an-illustrated-overview-04d193d8fec6?gi=6a258d737c25
https://winder.ai/llm-architecture-rag-implementation-design-patterns/
https://arxiv.org/pdf/2005.11401
https://arxiv.org/html/2403.19889v1
https://towardsdatascience.com/the-llm-triangle-principles-to-architect-reliable-ai-apps-d3753dd8542e?gi=4a84b1d08ecf
https://gist.github.com/Donavan/62e238aa0a40ca88191255a070e356a2
https://towardsdatascience.com/17-advanced-rag-techniques-to-turn-your-rag-app-prototype-into-a-production-ready-solution-5a048e36cdc8
https://medium.com/@cdg2718/why-your-rag-doesnt-work-9755726dd1e9
https://ai.gopubby.com/advanced-rag-retrieval-strategies-flow-and-modular-672493acb4a7?gi=12bd4ebef541
https://about.xethub.com/blog/you-dont-need-a-vector-database
	https://github.com/xetdata/RagIRBench/blob/main/you_do_not_need_a_vector_database.ipynb
	https://github.com/xetdata/RagIRBench
https://medium.com/decodingml/the-4-advanced-rag-algorithms-you-must-know-to-implement-5d0c7f1199d2
https://blog.demir.io/advanced-rag-implementing-advanced-techniques-to-enhance-retrieval-augmented-generation-systems-0e07301e46f4?gi=d5bbaa1b0d23
https://www.youtube.com/watch?v=Zj5RCweUHIk
https://blog.gopenai.com/mastering-rag-chunking-techniques-for-enhanced-document-processing-8d5fd88f6b72?gi=5c83cb32b55b
https://medium.com/@samarrana407/mastering-rag-advanced-methods-to-enhance-retrieval-augmented-generation-4b611f6ca99a
https://blog.demir.io/advanced-rag-implementing-advanced-techniques-to-enhance-retrieval-augmented-generation-systems-0e07301e46f4?gi=0afd4fce4b24
https://ai.gopubby.com/advanced-rag-12-enhancing-global-understanding-b13dc9a8db39?gi=9591e17edd13
https://huggingface.co/blog/hrishioa/retrieval-augmented-generation-1-basics
https://huggingface.co/blog/hrishioa/retrieval-augmented-generation-2-walking
https://huggingface.co/blog/hrishioa/retrieval-augmented-generation-3-structure
https://www.grey-wing.com/blog/how-we-had-to-build-ocean-oracle

**Comment by rmusser01** [comment]
https://github.com/whitead/paper-qa
https://blog.dagworks.io/p/building-a-conversational-graphdb
https://github.com/pymupdf/RAG
https://medium.com/microsoftazure/introducing-graphrag-with-langchain-and-neo4j-90446df17c1e
https://medium.com/@Stan_DS/build-a-100-free-hallucination-free-secure-rag-chatbot-using-reranker-and-gpt4o-96c2eea24f95
https://blog.demir.io/advanced-rag-implementing-advanced-techniques-to-enhance-retrieval-augmented-generation-systems-0e07301e46f4?gi=9e3e04c616da
https://superlinked.com/vectorhub/articles/advanced-retrieval-augmented-generation
https://div.beehiiv.com/p/advanced-rag-series-retrieval
https://towardsdatascience.com/how-to-reduce-embedding-size-and-increase-rag-retrieval-speed-7f903d3cecf7?gi=b5bcdb8dfecf
https://generativeai.pub/advanced-rag-retrieval-strategy-query-rewriting-a1dd61815ff0
https://medium.com/intel-tech/tabular-data-rag-llms-improve-results-through-data-table-prompting-bcb42678914b
https://generativeai.pub/advanced-rag-retrieval-strategies-hybrid-retrieval-997d39659720?gi=09c4b7d4c4c6
https://ai.gopubby.com/advanced-rag-retrieval-strategy-embedded-tables-fdb3e44003a5?gi=59a21f193fb7

**Comment by rmusser01** [comment]
Eval Rag

https://docs.ragas.io/en/latest/index.html

**Comment by rmusser01** [comment]
Moving comment to this issue for tracking:

A comment chain from reddit that seems relevant:
```
cyan2k
My experience over the past 2-ish years:

RAG IS FUCKING DIFFICULT.

We're talking about a client shelling out >$200K for a solution that barely scrapes a 60% hit rate—the 
kind that would spark outrage and spawn strongly worded Reddit threads if you knew just how much it sucks.

So, set your expectations accordingly. You're not getting your perfect RAG in an afternoon.

"What do you mean difficult? It's three lines of code with LangChain, lol, noob," Sure, getting started is 
simple but making it actually good? lol right back at you.

You see plenty of people disappointed with RAG, claiming it doesn't meet business needs and isn't as 
great as hyped. But in my opinion, it’s more a case of "people don't have a fucking clue." They put together 
the most overengineered data transformation pipelines possible, using the latest "meta" they picked up from
 Medium or TikTok, all held together by the shittiest code lifted straight from LangChain and LlamaIndex. 
Then they head over to [r/MachineLearning](https://www.reddit.com/r/MachineLearning/) to complain about 
how RAG sucks, totally overlooking that your latest paper on some obscure 
RAG optimization means shit in the real world with real data.

Or they opt for the lazy approach... Just chunk it up, rely on large context windows, dump everything into a
 single vector store, and trust in the magic of the LLM to somehow make the result good. But then reality 
hits when it hallucinates the shit out over the 12,000 tokens you fed it. "Needle in a haystack," my ass. 
This approach has also zero real-world applicability.

So yeah, it's tough, but I promise, you can make any use case work. But getting there is real work, and you 
need to decide if the effort is worth it.

My Recommendations:

    Set Up an Evaluation Pipeline Early: You need to know how good your system is. Don’t start any real work 
without the ability to quantify performance. You must be able to say, "If I switch this prompt for that one, 
performance improves by 7% but hallucination increases by 10%."

    Educate Yourself: Dive into resources like ragas, dspy, and probably also libraries providing 
structured output guardrails like outlines.

    Choose Your Tracing/Logging Framework: Make friends with a framework like W&B, Langsmith, Langfuse, 
and my personal favorite, Phoenix.

    ragas docs: https://docs.ragas.io/en/stable/

    DSP< GitHub repository: https://github.com/stanfordnlp/dspy

    Outlines GitHub repository: https://github.com/outlines-dev/outlines

    wandb site: https://wandb.ai/site

    Langsmith: https://www.langchain.com/langsmith

    Phoenix: https://phoenix.arize.com/

Now you have the tools to really measure and optimize your system!

This is mostly trial and error, as dumb as it sounds. The tech is still too new to have a solid
 "best practice" guide. Data varies so much from Source A to Source B that the strategy has to be 
tailored for each client, for each use case. However, some ideas from a couple of successful projects:

if you need over 8k tokens, your chunking strategy, retrieval process, ranking, or whatever, SUCKS. 
That's why it blows my mind every time I hear people complain that Llama3 only has an 8k token 
context. What do you even need more tokens for? What kind of magical text do you have that is 
so informationally dense over 5000 words that you can't split it?

8k is optimal because with most models (even those crazy high-context window ones like Yi), the 
first 8k tokens are amazingly accurate, and they ALL drop off. Perhaps not in your "passphrase" 
search test, but in real life with real end-users, they will.

Less Is More: Don’t get caught up in merging 20 different RAG pipelines from LangChain and 30
 from LlamaIndex all put together into one frankenstein monster of a pipeline just because you 
see in the docs that it's a direct implementation of some new paper. Stick to what works and use 
your own brain. Financial data, for example, often just boils down to a few key numbers, but 
understanding them requires a web of related data across documents. Classic RAG will let you 
down here. You need a system that maps entities via relationships like a Knowledge Graph.

The paper you read yesterday about RAPTOR, HYDE or some new "hot shit" RAG stuff... yeah try it out, but keep
 in mind they are usually not tested with real use cases. Alone the fact that you as the study creator 
know that you only test on benchmarks makes your whole study biased imho, and 90% of those papers 
don't publish a single line of code anyway, which is fucking stupid, and should be prohibited, because who
 the fuck knows if the numbers they put into their study are even true. Don't be desillusionized if it don't
 work out. Also never go to [r/machinelearning](https://www.reddit.com/r/machinelearning/) for your 
LLM questions. Saddest place on reddit lol.

Have Stamina! Good Luck!

TL;DR: Set up an evaluation pipeline, improve your process step by step. Keep it simple and 
chill with context size and optimization craziness.


----->
arthurwolf

    Just chunk it up, rely on large context windows, dump everything into a single vector store,
 and trust in the magic of the LLM to somehow make the result good. But then reality hits 
when it hallucinates the shit out over the 12,000 tokens you fed it

The solution we implemented is similar to this but with an extra step.

We gather data *very* liberally (using both a keyword and a vector based search), get anything
 that might be related. Massive amounts of tokens.

Then we go over each result, and for each result, we ask it « is there anything in there that 
matters to this question? <question>. if so, tell us what it is ».

Then with only the info that passed through that filter, we do the actual final prompt as 
you'd normally do (at that point we are back down to pretty low numbers of tokens).

Got us from around 60% to a bit over 85%, and growing (which is fine for our use case).

It's pretty fast (the filter step is highly parralelizable), and it works for *most* requests 
(but fails miserably for a few, something for which we're implementing contingencies).

However, it is expensive. Talking multiple cents per customer question. That might not
 be ok for others. We are exploring using (much) cheaper models for the filter and seeing good results so far.

----->----->
nightman

I recommend to try Reranking (like Cohere reranking and filtering based on relevance_score) 
instead of current filtering. It might not work for you but it's a middle ground between naive
 vector store retreival and checking each document with LLM if it fits.
```

**Comment by rmusser01** [comment]
LongRAG
https://github.com/TIGER-AI-Lab/LongRAG

**Comment by rmusser01** [comment]
Add Wikipedia + X wiki support.

**Comment by rmusser01** [comment]
Un-examined
ColBERT
	https://www.answer.ai/posts/2024-08-13-small-but-mighty-colbert.html
	https://arxiv.org/abs/2406.11706
	https://arxiv.org/abs/2303.00807

RAG Papers
	https://arxiv.org/abs/2408.05025
	https://arxiv.org/pdf/2408.02545
	https://arxiv.org/pdf/2408.03910
	https://arxiv.org/pdf/2408.03325
	https://arxiv.org/abs/2407.19813
	https://arxiv.org/abs/2407.19825
	https://arxiv.org/abs/2407.21712
	https://arxiv.org/abs/2407.19825
	https://arxiv.org/abs/2407.13481
	https://arxiv.org/abs/2408.02545
https://arxiv.org/abs/2408.06778
	https://arxiv.org/abs/2408.03910

RAG Techniques Implemented
	https://github.com/ShishirPatil/gorilla/tree/main/raft
	https://arxiv.org/html/2408.06941v1
	https://github.com/NirDiamant/RAG_Techniques
	https://github.com/NirDiamant/Controllable-RAG-Agent
	https://github.com/EdwardDali/erag
	https://github.com/gusye1234/nano-graphrag

Graph Reasoning
	https://jdsemrau.substack.com/p/graph-reasoning-structures-and-concepts

Eval
	https://arxiv.org/pdf/2408.02666v1
	https://arxiv.org/abs/2407.18416
	https://arxiv.org/html/2408.02666v1

Reasoning
	https://arxiv.org/abs/2408.06195

**Comment by rmusser01** [comment]
Well, I aint waiting 6 months.
RAG
```
Systematically Improving RAG Applications prep:

**Week 1: Cold Start Problem**

- Overview of the Playbook and RAG System Inference Flywheel
- Understanding and Justifying the Playbook
- LLM Synthetic Data Generation for Evaluation
- Leading vs Lagging Metrics
- Evaluation and Improvement Strategies

**Week 2: Identifying Failure Modes**

- Segmenting input space of questions
- Developing specialized systems based on topics and  capabilities 
- Improving precision and recall before considering generation
- Working with domain experts to decompose problem space 

**Week 3: Extending Capabilities**

- Understanding limited inventory and capabilities issues 
- Metadata filtering
- Rerankers w/ COhere 
- Fine-Tuning Embedding Models
- Multimodal data handling (PDFs, PPTs, Tables, Images)

**Week 4: Representations**

- Understanding Embeddings and Search
- Routing queries using classifiers
- Evaluating routing effectiveness
- Handling multiple routes
- Parallel Routing with instructor 
- Identifying what tools to provide agents 

**Week 5: Product**

- Generation techniques
- Streaming responses
- Managing citations
- Feedback UI and collecting labels
- Prompt engineering and rejecting wor
```

**Comment by rmusser01** [comment]
Fuck it. we'll do it live.

https://github.com/asg017/sqlite-lembed
https://github.com/asg017/sqlite-vec

**Comment by rmusser01** [comment]
Tracking RAG-related research

https://arxiv.org/pdf/2408.14906
https://arxiv.org/pdf/2407.14482
https://www.turingpost.com/p/longrag
https://arxiv.org/pdf/2408.14717

**Comment by rmusser01** [comment]
https://pub.towardsai.net/reliable-agentic-rag-with-llm-trustworthiness-estimates-c488fb1bd116
https://docs.datacommons.org/papers/DataGemma-FullPaper.pdf

**Comment by rmusser01** [comment]
https://huggingface.co/lightblue/kurage-multilingual

**Comment by rmusser01** [comment]
Closing since RAG has been added. And all links have been shuffled/sorted to their proper issues page.
